### PR TITLE
[COR-86] Remove Simple Queries

### PR DESCRIPTION
**Pull request by goedderz**
### Scope & Purpose

Enterprise part: https://github.com/arangodb/enterprise/pull/1601

---

> [!NOTE]
> **Medium Risk**
> This is a breaking API removal affecting REST clients and arangosh/JS scripts that relied on `/_api/simple/*` and `@arangodb/simple-query`. Risk is moderated by being largely deletion-focused, but any remaining callers will now fail at runtime and the new AQL-based helpers may differ subtly in semantics/performance.
> 
> **Overview**
> **Removes the Simple Query REST API surface** by deleting server handlers and route registration for `/_api/simple/*` endpoints (e.g. `all`, `all-keys`, `by-example`, `lookup-by-keys`, `remove-by-keys`) and dropping the associated `RestSimpleHandler`/`RestSimpleQueryHandler` sources and path constants.
> 
> **Cleans up and reshapes JS client APIs** by deleting `@arangodb/simple-query` and `simple-query-common`, removing collection methods that depended on them (including geo helpers), and reimplementing a retained subset (`any`, `all`, `range`, `byExample`, `firstExample`, `removeByExample`, `replaceByExample`, `updateByExample`) directly via AQL (with reduced support such as no `skip`/`limit` for some helpers). Tests are updated to stop exercising removed endpoints and to use document APIs instead (including switching batch deletes to `DELETE /_api/document/...`).
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 244837c5a5381c4fad8c1622f63a69c1b356bd01. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

### CODE REVIEW

File: arangod/RestHandler/RestDocumentHandler.cpp [hunk 1]
```diff
@@ -845,6 +845,17 @@ async<void> RestDocumentHandler::readManyDocuments() {
...
+  // Add GLOBAL_MANAGED hint for smart edge collections
+  if (ServerState::instance()->isCoordinator()) {
+    CollectionNameResolver resolver{_vocbase};
+    auto col = resolver.getCollection(cname);
+    if (col != nullptr && col->isSmartEdgeCollection()) {
+      // Smart Edge Collections hit multiple shards with dependent requests,
+      // they have to be globally managed.
+      trx->addHint(transaction::Hints::Hint::GLOBAL_MANAGED);
+    }
+  }
```
Comment (goedderz) [inline arangod/RestHandler/RestDocumentHandler.cpp]: Interesting, was this trx hint missing here? Good catch! But I think this change should go into a separate PR, it gets lost here. Is it covered by the tests (in this PR)? And I think you could call `addTransactionHints()` instead, it contains the same code.

File: arangod/GeneralServer/GeneralServerFeature.cpp [hunk 2]
```diff
@@ -640,30 +638,13 @@ void GeneralServerFeature::defineRemainingHandlers(
...
-      RestVocbaseBaseHandler::SIMPLE_REMOVE_PATH,
-      RestHandlerCreator<RestSimpleHandler>::createData<aql::QueryRegistry*>,
-      queryRegistry);
+#ifdef USE_V8
+  if (server().isEnabled<V8DealerFeature>()) {
+    // the tasks feature depends on V8. only enable it if JavaScript is enabled
+    f.addPrefixHandler(RestVocbaseBaseHandler::TASKS_PATH,
+                       RestHandlerCreator<RestTasksHandler>::createNoData);
+  }
+#endif
```
Comment (mpoeter) [inline arangod/GeneralServer/GeneralServerFeature.cpp]: Since this is for 4.0 where we no longer have V8 this should be removed.

File: tests/js/client/server_parameters/test-streaming-max-transaction-size.js [hunk 3]
```diff
@@ -162,10 +162,10 @@ function testSuite() {
       let docid = `${cn}/500`;
       let trx = db._createTransaction({ collections: { write: [c] } });
       trx.query(aql`LET d = DOCUMENT(${docid})
-                        UPDATE d WITH { "aaa": "aaa" } IN ${c}
+                        UPDATE d WITH { "aaa": "aaa" } IN ${aql.literal(cn)}
```
Comment (mpoeter) [inline tests/js/client/server_parameters/test-streaming-max-transaction-size.js:165]:
```suggestion
                        UPDATE d WITH { "aaa": "aaa" } IN ${cn}
```

File: js/actions/api-simple.js
Comment (mpoeter) [inline js/actions/api-simple.js:1]: This file should be completely removed.

File: arangod/RestHandler/RestDocumentHandler.cpp (hunk 1 above)
Comment (mpoeter) [inline arangod/RestHandler/RestDocumentHandler.cpp]: Why was this added?

File: js/client/modules/@arangodb/simple-query.js
Comment (mpoeter) [inline js/client/modules/@arangodb/simple-query.js:1]: Check if all the changes in this file are necessary.

File: arangod/GeneralServer/GeneralServerFeature.cpp (hunk 2 above)
Comment (bluepal-avanthi-dundigala) [reply arangod/GeneralServer/GeneralServerFeature.cpp]: Removed V8 code.

File: js/actions/api-simple.js
Comment (bluepal-avanthi-dundigala) [reply js/actions/api-simple.js:1]: Removed this file.

File: arangod/RestHandler/RestDocumentHandler.cpp (hunk 1 above)
Comment (bluepal-avanthi-dundigala) [reply arangod/RestHandler/RestDocumentHandler.cpp]: These lines were added to fix a failing test when reading from smart edge collections using the document API. Smart edge collections span multiple shards, and without the GLOBAL_MANAGED hint, transactions were not properly coordinated across shard leaders, causing failures in cluster mode. Adding GLOBAL_MANAGED ensures the transaction starts on all relevant shards, enabling correct and consistent reads.

File: js/client/modules/@arangodb/simple-query.js
Comment (bluepal-avanthi-dundigala) [reply js/client/modules/@arangodb/simple-query.js:1]: All changes in js/client/modules/@arangodb/simple-query.js are necessary because:
- The JavaScript API (collection.byExample(), etc) is not being deprecated
- The FILTER keyword fix resolves a bug that generates invalid AQL
- The defensive changes align with patterns in arango-collection.js and improve robustness
- These methods are still used in production code (Foxx modules, etc)

File: js/client/modules/@arangodb/testutils/collect-crud-api-generic-tests.js [hunk 4]
```diff
@@ -205,6 +207,10 @@ const generateTestSuite = (collectionWrapper, testNamePostfix = "") => {
         if (isEnterpriseGraphEdge) {
           // Keys are not provided by us, they are auto-generated.
           keys = result.parsedBody.map(d => d._key);
+          // Update documents array with generated keys for later comparison
+          for (let i = 0; i < documents.length; ++i) {
+            documents[i]._key = keys[i];
+          }
```
Comment (mpoeter) [inline js/client/modules/@arangodb/testutils/collect-crud-api-generic-tests.js]: Why is this necessary? The code above generates documents with predefined keys.

File: js/client/modules/@arangodb/testutils/collect-crud-api-generic-tests.js [hunk 5]
```diff
@@ -219,62 +225,80 @@ const generateTestSuite = (collectionWrapper, testNamePostfix = "") => {
...
+        // For Enterprise Graph edges in single-server mode, use keys directly
+        // For Enterprise Graph Smart vertex collections in cluster mode, use keys directly
+        // because the sharding strategy requires the key to determine the responsible shard
+        const isCluster = internal.isCluster();
+        const isEnterpriseGraphSmartVertex = collectionWrapper._isEnterpriseGraphSmartVertex;
+        const lookupData = ((isEnterpriseGraphEdge && !isCluster) || (isEnterpriseGraphSmartVertex && isCluster))
+          ? keys 
+          : keys.map(key => collection.name() + '/' + key);
+        const lookupUrl = '/_api/document/' + encodeURIComponent(collection.name()) + '?onlyget=true';
+        const result = arango.PUT_RAW(lookupUrl, lookupData);
```
Comment (mpoeter) [inline js/client/modules/@arangodb/testutils/collect-crud-api-generic-tests.js]: I don't understand this comment. Also, this should be about looking up documents, so this should be a GET request, not a PUT.

At closer look, I think the whole `testLookupByKeys` test should be removed. The comment for the tests says "this tests the SimpleQueries lookupByKeys and removeByKeys", but we have removed those simple queries, so there is no point in adjusting the test.

File: js/client/modules/@arangodb/arango-collection.js
Comment (mpoeter) [inline js/client/modules/@arangodb/arango-collection.js:1]: This file needs more work. The function prototypes should match the `else` part of the `if SYS_IS_V8_BUILD` from the old code. Right now some match the `then` part instead.

File: js/server/modules/@arangodb/simple-query.js
Comment (mpoeter) [inline js/server/modules/@arangodb/simple-query.js:1]: We should be able to remove this file since we no longer have server side JS.
In fact, we should be able to remove the whole `js/server` folder, but for this PR I would only remove this file.

File: js/client/modules/@arangodb/simple-query.js
Comment (mpoeter) [reply js/client/modules/@arangodb/simple-query.js:1]: I am not convinced that we really need this. Foxx is deprecated and will be removed.

Why did the test for non-V8 builds work _without_ these changes? Is the code actually used somewhere?

But even if we really need this file and have to make some adjustments, I think this code can be simplified a lot!

Comment (bluepal-avanthi-dundigala) [reply js/client/modules/@arangodb/simple-query.js:1]: Yes, the code is definitely used.
collection.byExample() and collection.firstExample() are used in test-extended-names-collections-on.js.
Both methods rely on SimpleQueryByExample, which calls SimpleQueryByExample.prototype.execute in simple-query.js.

This logic is used in client-side (arangosh), server-side, and common collection modules.

As for why the non-V8 tests passed without these changes:
The test file runs via arangosh, which requires V8.
The tests only use simple example objects like { value1: 36 }.
The bug only occurs when using dotted keys (e.g., { "a.b": 123 }) or special-character cases.
Since those scenarios are not covered by the test, the issue was never triggered.

So the code is used, but the specific edge case affected by this change is simply not tested, which is why the non-V8 builds appeared to work fine.

I agree it could be simplified, so it would help to understand which parts you think should be refactored or improved.

Comment (mpoeter) [reply js/client/modules/@arangodb/simple-query.js:1]: Yes, this runs on the client, but the old code still sent a request to the simple-query endpoint which should not be available in a non-V8 build, so I am wondering why that did not fail.

What happens if you revert the changes to this file? Which tests break then?
